judge/models/tests/test_contest.py

```diff
@@ -0,0 +1,454 @@
+                'can_see_full_scoreboard': self.assertTrue,
+                'is_accessible_by': self.assertTrue,
+                'is_editable_by': self.assertFalse,
+                'is_in_contest': self.assertFalse,
+            },
+            'anonymous': {
+                # False because contest has not begun
+                'can_see_own_scoreboard': self.assertFalse,
+                'can_see_full_scoreboard': self.assertFalse,
+                'is_accessible_by': self.assertFalse,
+                'is_editable_by': self.assertFalse,
+                'is_in_contest': self.assertFalse,
+            },
+        }
+        self._test_object_methods_with_users(self.organization_private_contest, data)
+
+    def test_private_user_contest_methods(self):
+        data = {
+            'superuser': {
+                'can_see_own_scoreboard': self.assertTrue,
+                'can_see_full_scoreboard': self.assertTrue,
+                'is_accessible_by': self.assertTrue,
+                'is_editable_by': self.assertTrue,
+                'is_in_contest': self.assertFalse,
+            },
+            'normal': {
+                'can_see_own_scoreboard': self.assertTrue,
+                'can_see_full_scoreboard': self.assertTrue,
+                'is_accessible_by': self.assertFalse,
+                'is_editable_by': self.assertFalse,
+                'is_in_contest': self.assertFalse,
+            },
+            'anonymous': {
+                'can_see_own_scoreboard': self.assertTrue,
+                'can_see_full_scoreboard': self.assertTrue,
+                'is_accessible_by': self.assertFalse,
+                'is_editable_by': self.assertFalse,
+                'is_in_contest': self.assertFalse,
+            },
+        }
+        self._test_object_methods_with_users(self.private_user_contest, data)
+
+    def test_contests_list(self):
+        for name, user in self.users.items():
+            with self.subTest(user=name):
+                # We only care about consistency between Contest.is_accessible_by and Contest.get_visible_contests
+                contest_keys = []
+                for contest in Contest.objects.prefetch_related('organizers', 'private_contestants', 'organizations'):
+                    if contest.is_accessible_by(user):
+                        contest_keys.append(contest.key)
+
+                self.assertListEqual(
+                    sorted(Contest.get_visible_contests(user).values_list('key', flat=True)),
+                    sorted(contest_keys),
+                )
+
+    def test_contest_clean(self):
+        _now = timezone.now()
+        contest = create_contest(
+            key='contest',
+            start_time=_now,
+            end_time=_now - timezone.timedelta(days=1),
+            problem_label_script='invalid',
+            format_config={'invalid': 'invalid'},
+        )
+        with self.assertRaisesRegex(ValidationError, 'ended before it starts'):
```

Comment:
Perhaps add a second test to make sure a contest can't end and start at the same time.

Reply:
Done